### PR TITLE
Use vectorized SpanHelpers.SequenceEqual for string equality

### PR DESCRIPTION
**Pull request by jkotas**
SpanHelpers.SequenceEqual is our optimized workhorse for comparing blocks of memory. Use it for string equality instead of ad-hoc optimized loops.

`"Hello world!".Equals("Hello world?")` is 20% faster on x64 with this change and it only gets better from there.

### CODE REVIEW

**Comment by jkotas** [comment]
Depends on https://github.com/dotnet/corefx/pull/28080 that is changing SpanHelpers.SequenceEqual argument to nuint.

**Comment by jkotas** [comment]
cc @marek-safar

File: src/mscorlib/src/System/String.Comparison.cs [hunk 1]
```diff
@@ -1126,7 +1018,10 @@ public Boolean StartsWith(String value, StringComparison comparisonType)
                     }
                     return (value.Length == 1) ?
                             true :                 // First char is the same and thats all there is to compare
-                            StartsWithOrdinalHelper(this, value);
+                            SpanHelpers.SequenceEqual(
```
Comment (ahsonkhan) [inline src/mscorlib/src/System/String.Comparison.cs]: What happens if this.Length is less than value.Length?

Comment (jkotas) [reply src/mscorlib/src/System/String.Comparison.cs]: This condition is checked two lines above this one.

**Comment by marek-safar** [comment]
@jkotas very nice

We are still using https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/String.Comparison.cs but this file looks quite easy to unify/share

File: src/mscorlib/src/System/String.Comparison.cs [hunk 2]
```diff
@@ -80,61 +80,17 @@ private static unsafe int CompareOrdinalIgnoreCaseHelper(String strA, String str
...
-                    if (*(int*)a != *(int*)b) return false;
-                    length -= 2; a += 2; b += 2;
-                }
-
-                return true;
-            }
+            return SpanHelpers.SequenceEqual(
+                    ref Unsafe.As<char, byte>(ref strA.GetRawStringData()),
+                    ref Unsafe.As<char, byte>(ref strB.GetRawStringData()),
+                    ((nuint)strA.Length) * 2);
```
Comment (kasper3) [inline src/mscorlib/src/System/String.Comparison.cs]: Maybe this `nunit` cast is not required?

>     4>src/System/String.Comparison.cs(93,21): error CS1503: Argument 3: cannot convert from 'ulong' to 'int' [/mnt/resource/j/workspace/dotnet_coreclr/master/x64_debug_centos7.1_innerloop_prtest/src/mscorlib/System.Private.CoreLib.csproj]
>     4>src/System/String.Comparison.cs(1024,33): error CS1503: Argument 3: cannot convert from 'ulong' to 'int' [/mnt/resource/j/workspace/dotnet_coreclr/master/x64_debug_centos7.1_innerloop_prtest/src/mscorlib/System.Private.CoreLib.csproj]

Comment (ahsonkhan) [reply src/mscorlib/src/System/String.Comparison.cs]: I believe it is. Waiting for the mirror to merge - https://github.com/dotnet/coreclr/pull/17001

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
@dotnet-bot test Ubuntu x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Pull request merged**